File: Nick-Haskell/PokerPuzzle.hsproj/PokerPuzzle.hs
{-# LANGUAGE MonadComprehensions #-}

module PokerPuzzle where
  
import Data.List (sortBy, group, find, nub)
import Data.Maybe (isJust, fromJust)
import Control.Monad (guard)
import Model

type Kickers = [Card]
type SortedHand = Hand
type SortedCards = [Card]

data PokerResult = 
    HighCard Card Kickers
  | OnePair (Card, Card) Kickers
  | TwoPairs (Card, Card) (Card, Card) Kickers
  | ThreeOfAKind (Card, Card, Card) Kickers
  | Straight (Card, Card, Card, Card, Card)
  | Flush (Card, Card, Card, Card, Card)
  | FullHouse (Card, Card, Card) (Card, Card)
  | FourOfAKind (Card, Card, Card, Card) Kickers
  | StraightFlush (Card, Card, Card, Card, Card)
  | RoyalFlush (Card, Card, Card, Card, Card)
  deriving (Show, Eq, Ord)

--

newtype GroupedHand = GroupedHand {groups :: [[Card]]}

--

allSameSuit :: [Card] -> Bool
allSameSuit cards = 
  let
    firstSuit = suit $ head cards
  in 
    all (\card -> suit card == firstSuit) cards
    
-- alternative implementation
allSameSuit2 :: [Card] -> Bool
allSameSuit2 cards = length (nub $ map suit cards) == 1 

 
isStraight :: SortedCards -> Bool
isStraight cards  = 
  let
    values = map value cards
    currentPreviousList = zip values $ tail values
    currentSuccessorPrevious (prev, current) = prev == pred current
  in 
    all currentSuccessorPrevious currentPreviousList 
   
--
 
cardsTuple2 [card1, card2] = (card1, card2)
cardsTuple3 [card1, card2, card3] = (card1, card2, card3) 
cardsTuple4 [card1, card2, card3, card4] = (card1, card2, card3, card4)
cardsTuple5 [card1, card2, card3, card4, card5] = (card1, card2, card3, card4, card5)

--
    
highCard :: SortedHand -> PokerResult
highCard (Hand cards) = HighCard (head cards) (tail cards) 

onePair :: GroupedHand -> Maybe PokerResult
onePair (GroupedHand groups) = 
  -- using monadic do, see threeOfAKind for applicative alternative
  do 
    pair <- find ((==2) . length) groups
    let remainingCards = concat $ filter (/=pair) groups
    return $ OnePair (cardsTuple2 pair) remainingCards
    
twoPairs :: GroupedHand -> Maybe PokerResult
twoPairs (GroupedHand groups) = 
  let
    allTwos = filter ((==2) . length) groups
    pair1 = cardsTuple2 $ allTwos !! 0
    pair2 = cardsTuple2 $ allTwos !! 1
    otherCard = head $ filter ((==1) . length) groups
  in
    -- if then else, formulation; alternative using `guard` below.  
    if (length allTwos == 2) then Just (TwoPairs pair1 pair2 otherCard) else Nothing
      
threeOfAKind :: GroupedHand -> Maybe PokerResult
threeOfAKind (GroupedHand groups) =
  let 
    triplet = cardsTuple3 <$> find ((==3) . length) groups
    remainingCards = concat $ filter ((/=3) . length) groups
  in
    ThreeOfAKind <$> triplet <*> pure remainingCards  
      
straight :: SortedHand -> Maybe PokerResult
straight (Hand cards) = 
  (Straight $ cardsTuple5 cards) <$ guard (isStraight cards)
    
flush :: Hand -> Maybe PokerResult
flush (Hand cards) = (Flush $ cardsTuple5 cards) <$ guard (allSameSuit cards)
    
fullHouse :: GroupedHand -> Maybe PokerResult
fullHouse (GroupedHand groups) = 
  let
    threeOfAKind = cardsTuple3 <$> find ((==3) . length) groups
    twoOfAKind = cardsTuple2 <$> find ((==2) . length) groups
  in
    FullHouse <$> threeOfAKind <*> twoOfAKind

fourOfAKind :: GroupedHand -> Maybe PokerResult
fourOfAKind (GroupedHand groups) = 
  let 
    quadruplet = cardsTuple4 <$> find ((==4) . length) groups
    otherCard = head $ filter ((==1) . length) groups
  in
    FourOfAKind <$> quadruplet <*> pure otherCard

straightFlush :: SortedHand -> Maybe PokerResult
straightFlush (Hand cards) =
  let
     isStraightFlush = allSameSuit cards && isStraight cards
  in
    (StraightFlush $ cardsTuple5 cards) <$ guard isStraightFlush
    
royalFlush :: SortedHand -> Maybe PokerResult
royalFlush (Hand cards) =
  let
    lowestCardValue = value $ head cards
    isRoyalFlush = allSameSuit cards && isStraight cards && lowestCardValue == Jack
  in
    -- monad comprehension version of `<$ guard` and if..then..else
    [RoyalFlush $ cardsTuple5 cards | isRoyalFlush]
    
--

pokerResult :: Hand -> PokerResult
pokerResult (Hand cardsInHand) = 
  let
    sortedCards = (sortBy (flip compare))  cardsInHand
    sortedHand = Hand sortedCards
    groupedHand = GroupedHand $ group sortedCards
    
    options = [
        royalFlush sortedHand
      , straightFlush sortedHand
      , fourOfAKind groupedHand
      , fullHouse groupedHand
      , flush sortedHand
      , straight sortedHand
      , threeOfAKind groupedHand
      , twoPairs groupedHand
      , onePair groupedHand
      , Just $ highCard sortedHand
      ]
      
    best = find isJust options
  in
    fromJust $ fromJust best
    
    
isPlayer1Winner :: (Hand, Hand) -> Bool
isPlayer1Winner (player1Hand, player2Hand) =
  let
     player1Result = pokerResult player1Hand
     player2Result = pokerResult player2Hand
  in
    player1Result > player2Result

File: Nick-Swift/PokerPuzzle/PokerPuzzle.swift
import Foundation

typealias Kickers = [Card]

enum PokerResult {
    case highCard(Card, Kickers)
    case onePair((Card, Card), Kickers)
    case twoPairs((Card, Card), (Card, Card), Kickers)
    case threeOfAKind((Card, Card, Card), Kickers)
    case straight((Card, Card, Card, Card, Card))
    case flush((Card, Card, Card, Card, Card))
    case fullHouse((Card, Card, Card), (Card, Card))
    case fourOfAKind((Card, Card, Card, Card), Kickers)
    case straightFlush((Card, Card, Card, Card, Card))
    case royalFlush((Card, Card, Card, Card, Card))
}

extension PokerResult {
    var rankValue: Int {
        let value:Int
        
        switch (self) {
        case .highCard:
            value = 0
        case .onePair:
            value = 1
        case .twoPairs:
            value = 2
        case .threeOfAKind:
            value = 3
        case .straight:
            value = 4
        case .flush:
            value = 5
        case .fullHouse:
            value = 6
        case .fourOfAKind:
            value = 7
        case .straightFlush:
            value = 8
        case .royalFlush:
            value = 9
        }
        return value
    }
}

// MARK: -
extension PokerResult : Equatable {
    static func == (lhs: PokerResult, rhs: PokerResult) -> Bool {
        switch (lhs, rhs) {
        case (let .highCard(card1, kickers1), let .highCard(card2, kickers2)):
            return card1 == card2 && kickers1 == kickers2
            
        case (let .onePair(cards1, kickers1), let .onePair(cards2, kickers2)):
            return cards1 == cards2 && kickers1 == kickers2
            
        case (let .twoPairs(cards1a, cards1b, kickers1), let .twoPairs(cards2a, cards2b, kickers2)):
            return cards1a == cards2a && cards1b == cards2b && kickers1 == kickers2
            
        case (let .threeOfAKind(cards1, kickers1), let .threeOfAKind(cards2, kickers2)):
            return cards1 == cards2 && kickers1 == kickers2
            
        case (let .straight(cards1), let .straight(cards2)):
            return cards1 == cards2
            
        case (let .flush(cards1), let .flush(cards2)):
            return cards1 == cards2
            
        case (let .fullHouse(cards1a, cards1b), let .fullHouse(cards2a, cards2b)):
            return cards1a == cards2a && cards1b == cards2b
            
        case (let .fourOfAKind(cards1, kickers1), let .fourOfAKind(cards2, kickers2)):
            return cards1 == cards2 && kickers1 == kickers2
            
        case (let .straightFlush(cards1), let .straightFlush(cards2)):
            return cards1 == cards2
            
        case (let .royalFlush(cards1), let .royalFlush(cards2)):
            return cards1 == cards2
            
        default:
            return false
        }
    }
}

func < (lhs: Array<Card>, rhs: Array<Card>) -> Bool {
    for i in 0...(lhs.count - 1) {
        if lhs[i] != rhs[i] {
            return lhs[i] < rhs[i]
        }
    }
    return false
}


extension PokerResult : Comparable {
    static func < (lhs: PokerResult, rhs: PokerResult) -> Bool {
        switch (lhs, rhs) {
        case (let .highCard(card1, kickers1), let .highCard(card2, kickers2)):
            if card1 != card2 {
                return card1 < card2
            }
            return kickers1 < kickers2
            
        case (let .onePair(cards1, kickers1), let .onePair(cards2, kickers2)):
            if cards1 != cards2 {
                return cards1 < cards2
            }
            return kickers1 < kickers2
            
        case (let .twoPairs(cards1a, cards1b, kickers1), let .twoPairs(cards2a, cards2b, kickers2)):
            if cards1a != cards2a {
                return cards1a < cards2a
            } else if cards1b != cards2b {
                return cards1b < cards2b
            }
            return kickers1 < kickers2
            
        case (let .threeOfAKind(cards1, kickers1), let .threeOfAKind(cards2, kickers2)):
            if cards1 != cards2 {
                return cards1 < cards2
            }
            return kickers1 < kickers2
            
        case (let .straight(cards1), let .straight(cards2)):
            return cards1 < cards2
            
        case (let .flush(cards1), let .flush(cards2)):
            return cards1 < cards2
            
        case (let .fullHouse(cards1a, cards1b), let .fullHouse(cards2a, cards2b)):
            if cards1a != cards2a {
                return cards1a < cards2a
            }
            return cards1b < cards2b
            
        case (let .fourOfAKind(cards1, kickers1), let .fourOfAKind(cards2, kickers2)):
            if cards1 != cards2 {
                return cards1 < cards2
            }
            return kickers1 < kickers2
            
        case (let .straightFlush(cards1), let .straightFlush(cards2)):
            return cards1 < cards2
            
        case (let .royalFlush(cards1), let .royalFlush(cards2)):
            return cards1 < cards2
            
        default:
            return lhs.rankValue < rhs.rankValue
        }
    }
}

// MARK: -
typealias GroupedHand = [[Card]]

// MARK: -

func allSameSuit(cards: [Card]) -> Bool {
    let firstSuit = cards.first!.suit
    let notSame = cards.first { $0.suit != firstSuit }
    return notSame == nil
}

func isStraight(cards : SortedCards) -> Bool {
    let lowestCardValue = cards.last!.value.rawValue
    let highestCardValue = cards.first!.value.rawValue
    
    for i in lowestCardValue...highestCardValue {
        if cards[i - lowestCardValue].value.rawValue != i {
            return false
        }
    }
    return true
}

// MARK: -

func cardsTuple2(array:SortedCards) -> (Card, Card) {
    return (array[0], array[1])
}

func cardsTuple3(array:SortedCards) -> (Card, Card, Card) {
    return (array[0], array[1], array[2])
}

func cardsTuple4(array:SortedCards) -> (Card, Card, Card, Card) {
    return (array[0], array[1], array[2], array[3])
}

func cardsTuple5(array:SortedCards) -> (Card, Card, Card, Card, Card) {
    return (array[0], array[1], array[2], array[3], array[4])
}

// MARK: -

func highCard(cards: SortedHand) -> PokerResult {
    return .highCard(cards.first!, Array(cards.dropFirst(1)))
}

func onePair(groups: GroupedHand) -> PokerResult? {
    guard let pair = groups.first(where: { $0.count == 2 }) else {
        return nil
    }
    
    let remainingCards = groups.filter { $0 != pair }
    return .onePair(cardsTuple2(array: pair), Array(remainingCards.joined()))
}

func twoPairs(groups: GroupedHand) -> PokerResult? {
    let allTwos = groups.filter { $0.count == 2 }
    if allTwos.count != 2 {
        return nil
    }
    
    let pair1 = cardsTuple2(array: allTwos[0])
    let pair2 = cardsTuple2(array: allTwos[1])
    let otherCard = groups.filter { $0.count == 1 }.first!
    
    return .twoPairs(pair1, pair2, otherCard)
}

func threeOfAKind(groups: GroupedHand) -> PokerResult? {
    guard let triplet = groups.first(where: { $0.count == 3 }) else {
        return nil
    }
    
    let remainingCards = groups.filter { $0.count != 3}.joined()
    return .threeOfAKind(cardsTuple3(array:triplet), Array(remainingCards))
}

func straight(cards: SortedHand) -> PokerResult? {
    guard isStraight(cards: cards) else {
        return nil
    }
    
    return .straight(cardsTuple5(array: cards))
}

func flush(cards: Hand) -> PokerResult? {
    guard allSameSuit(cards: cards) else {
        return nil
    }
    
    return .flush(cardsTuple5(array: cards))
}

func fullHouse(groups: GroupedHand) -> PokerResult? {
    guard let threeOfAKind = groups.first(where: { $0.count == 3 }), let twoOfAKind = groups.first(where: { $0.count == 2 }) else {
        return nil
    }
    return .fullHouse(cardsTuple3(array: threeOfAKind), cardsTuple2(array: twoOfAKind))
}

func fourOfAKind(groups: GroupedHand) -> PokerResult? {
    guard let quadruplet = groups.first(where: { $0.count == 4 }) else {
        return nil
    }
    let remainingCards = groups.filter { $0.count == 1 }.first!
    return .fourOfAKind(cardsTuple4(array:quadruplet), remainingCards)
}

func straightFlush(cards: SortedHand) -> PokerResult? {
    guard allSameSuit(cards: cards) && isStraight(cards: cards) else {
        return nil
    }
    return .straightFlush(cardsTuple5(array: cards))
}

func royalFlush(cards: SortedHand) -> PokerResult? {
    let lowestCardValue = cards.last!.value
    guard allSameSuit(cards: cards) && isStraight(cards: cards) && lowestCardValue == .jack else {
        return nil
    }
    return .royalFlush(cardsTuple5(array: cards))
}

// MARK: -

extension Array where Element == Card {
    func group() -> [[Card]] {
        var current = [Card]()
        var result = [[Card]]()
        for i in self {
            if current.isEmpty || current.first! == i {
                current.append(i)
            } else {
                result.append(current)
                current = [i]
            }
        }
        result.append(current)
        return result
    }
}


func pokerResult(cards:Hand) -> PokerResult {
    let sortedCards = cards.sorted { $0 > $1 }
    let groupedCards = sortedCards.group()
    
    if let result = royalFlush(cards: sortedCards) {
        return result
    } else if let result = straightFlush(cards: sortedCards) {
        return result
    } else if let result = fourOfAKind(groups: groupedCards) {
        return result
    } else if let result = fullHouse(groups: groupedCards) {
        return result
    } else if let result = flush(cards: sortedCards) {
        return result
    } else if let result = straight(cards: sortedCards) {
        return result
    } else if let result = threeOfAKind(groups: groupedCards) {
        return result
    } else if let result = twoPairs(groups: groupedCards) {
        return result
    } else if let result = onePair(groups: groupedCards) {
        return result
    } else {
        return highCard(cards: sortedCards)
    }
}

func isPlayer1Winner(handTuple: (Hand, Hand)) -> Bool {
    let player1Result = pokerResult(cards: handTuple.0)
    let player2Result = pokerResult(cards: handTuple.1)
    
    return player1Result > player2Result
}

